### PR TITLE
Add examples to CodeService.md docs

### PR DESCRIPTION
**Pull request by swwilson-bsi**

### CODE REVIEW

File: docs/learning/backend/CodeService.md [hunk 1]
```diff
@@ -125,12 +124,76 @@ Generally it is a good idea to reserve a group of codes together rather than one
...
+Here are the steps to follow.
+
+First, you must choose the code service that should be used at the time that you create your iModel. The identity of the code service and information about how to connect  are stored in the iModel persistently.
+
+When you open a BriefcaseDb for that iModel, it will read the configuration data from the iModel and automatically try to connect to the code service. To support that, apps that access the iModel must provide the client for the code service and set `CodeService.createForIModel` to point to it.
+
+For example, if you configure your new iModel to use Bentley's cloud-based code service, apps must provide the `ITwinCodeService` client. To do that, an app simply depends on the `@bentley/code-service` package and then imports it.
+
+```ts
+import { ITwinCodeService } from "@bentley/code-service";
```
Comment (calebmshafer) [inline docs/learning/backend/CodeService.md:136]: This is currently an internal package so can't be in our public facing docs. Can you take out the section talking about this as an example?

Comment (swwilson-bsi) [reply docs/learning/backend/CodeService.md]: I made the requested changes

File: docs/learning/backend/CodeService.md [hunk 2]
```diff
@@ -125,12 +124,64 @@ Generally it is a good idea to reserve a group of codes together rather than one
 - `reserveCode` to reserve a single code
 - `reserveCodes` to reserve a set of codes
 
+Here are the steps to follow.
+
+The identity of the code service and information about how to connect are stored in the iModel persistently. When you open a BriefcaseDb for that iModel, it will read the configuration data from the iModel and automatically try to connect to the code service. When an app opens a briefcase, the briefcase will connect to the code service using the connection information stored in the iModel. The briefcase `.codeService` property will refer to the connected code service client, and apps must use that to reserve, update, and query codes and CodeSpecs.
```
Comment (kabentley) [inline docs/learning/backend/CodeService.md:129]:
```suggestion
The identity of the code service and information about how to connect are stored in the iModel persistently. When you open a BriefcaseDb for that iModel, it will read the configuration data from the iModel and automatically try to connect to the code service. When an app opens a briefcase for editing, the briefcase will connect to the code service using the connection information stored in the iModel and the user's credentials. The briefcase `.codeService` property will refer to the connected CodeService client, and apps must use that to reserve, update, and query codes and CodeSpecs.
```

File: docs/learning/backend/CodeService.md [hunk 3]
```diff
@@ -125,12 +124,64 @@ Generally it is a good idea to reserve a group of codes together rather than one
...
+const iModel = await BriefcaseDb.open({ fileName: props.fileName });
+```
+
+``` ts
+[[include:CodeService.reserveInternalCodeForNewElement]]
+```
+
+The logic to reserve an external code is similar - just use the `externalCodes` property instead.
+
+Note that reserveCode is asynchronous. It locks and unlocks the code index in the cloud.
```
Comment (kabentley) [inline docs/learning/backend/CodeService.md:144]:
```suggestion
Note that reserveCode is asynchronous and must be `await`ed. It locks and then updates the code index in the cloud.
```

File: docs/learning/backend/CodeService.md [hunk 4]
```diff
@@ -125,12 +124,64 @@ Generally it is a good idea to reserve a group of codes together rather than one
...
+
+Note the use of the `CodeService.makeProposedCode` helper function.
+
+Note that the reserveCode function will throw if the code is already reserved. If that happens when you are trying to insert a new element, it means that you cannot use that code.
```
Comment (kabentley) [inline docs/learning/backend/CodeService.md:148]:
```suggestion
The `reserveCode` function will throw if the code is already reserved. If that happens when you are trying to insert a new element, it means that someone else has already reserved that code and you should choose a different one.
```

File: docs/learning/backend/CodeService.md [hunk 5]
```diff
@@ -125,12 +124,64 @@ Generally it is a good idea to reserve a group of codes together rather than one
...
 - `updateCodes` to modify the properties of a set of codes
 - `deleteCodes` to delete a set of codes
 
+For example,
+
+``` ts
+[[include:CodeService.updateInternalCodeForExistinglement]]
+```
+
+Note that the element's `federationGuid` must be used as the Code's GUID.
```
Comment (kabentley) [inline docs/learning/backend/CodeService.md:170]:
```suggestion
Note that the element's `federationGuid` is the Code's Guid.
```

File: example-code/snippets/src/backend/ExampleCode.test.ts [hunk 6]
```diff
@@ -201,6 +201,60 @@ describe("Example Code", () => {
...
 
+  it("CodeService", async () => {
+
+    if (false) { // this will compile but it will not run, because the root elementhas no federationGuid -- waiting for a fix
+
+      // __PUBLISH_EXTRACT_START__ CodeService.reserveInternalCodeForNewElement
+      const code = Subject.createCode(iModel, IModel.rootSubjectId, "test"); // example code
+
+      const proposedCode = CodeService.makeProposedCode({ iModel, code, props: { guid: Guid.createValue() } });
+      await iModel.codeService?.internalCodes?.writeLocker.reserveCode(proposedCode);
```
Comment (kabentley) [inline example-code/snippets/src/backend/ExampleCode.test.ts:212]: you should either add a try/catch here or note that this function will throw and not insert a new subject if someone else has already added a subject named "test" (probably a better name would help).

File: example-code/snippets/src/backend/ExampleCode.test.ts [hunk 7]
```diff
@@ -201,6 +201,60 @@ describe("Example Code", () => {
...
+
+      const elementId = Subject.insert(iModel, IModel.rootSubjectId, code.value);
+      // __PUBLISH_EXTRACT_END__
+
+      // __PUBLISH_EXTRACT_START__ CodeService.updateInternalCodeForExistinglement
+      const el = iModel.elements.getElement(elementId);
+      el.code = new Code({ ...el.code.toJSON(), value: "changed" });
+      await iModel.codeService?.internalCodes?.writeLocker.updateCode({ guid: el.federationGuid!, value: el.code.value });
```
Comment (kabentley) [inline example-code/snippets/src/backend/ExampleCode.test.ts:220]: same here with note that this will fail if "changed" is already used.

**Comment by kabentley** [comment]
The concept of "reserve code" is sometimes decoupled from inserting elements for external codes. People may reserve codes (e.g. a range of sequential codes) and then use them as needed. That should be noted in the external code section,

File: docs/learning/backend/CodeService.md (hunk 1 above)
Comment (kabentley) [reply docs/learning/backend/CodeService.md]: > How will the apps that our customers write deal with internal codes?

They will use iTS.

File: docs/learning/backend/CodeService.md (hunk 5 above)
Comment (swwilson-bsi) [reply docs/learning/backend/CodeService.md]: done

File: docs/learning/backend/CodeService.md (hunk 4 above)
Comment (swwilson-bsi) [reply docs/learning/backend/CodeService.md]: done

File: docs/learning/backend/CodeService.md (hunk 3 above)
Comment (swwilson-bsi) [reply docs/learning/backend/CodeService.md]: done

File: docs/learning/backend/CodeService.md (hunk 2 above)
Comment (swwilson-bsi) [reply docs/learning/backend/CodeService.md]: done

File: example-code/snippets/src/backend/ExampleCode.test.ts (hunk 7 above)
Comment (swwilson-bsi) [reply example-code/snippets/src/backend/ExampleCode.test.ts]: done

**Comment by swwilson-bsi** [comment]
Code samples and more docs for code service

File: core/generated-docs/core/ecsql-common/index.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (kabentley) [inline core/generated-docs/core/ecsql-common/index.ts:1]: why is this here?

Comment (swwilson-bsi) [reply core/generated-docs/core/ecsql-common/index.ts]: was probably generated by the `builddocsite` command that I ran locally. I thought it was needed, but I see that it couldn't be. I deleted the files.

**Review by kabentley — APPROVED** [review]

**Review by pmconne — APPROVED** [review]

**Review by calebmshafer — APPROVED** [review]
